### PR TITLE
Persist gossip contact info

### PR DESCRIPTION
**Pull request by mvines**
Cold starts for gossip can take a while, and node IP addresses really don't change often.  Save gossip contact info for faster validator restarts

### CODE REVIEW

File: core/src/cluster_info.rs [hunk 1]
```diff
@@ -653,6 +663,117 @@ impl ClusterInfo {
...
+                .unwrap()
+                .iter()
+                .map(|contact_info| contact_info.gossip)
+                .collect::<HashSet<_>>();
+
+            gossip
+                .crds
+                .get_nodes()
+                .filter_map(|v| {
+                    if let Some(contact_info) = v.value.contact_info() {
```
Comment (behzadnouri) [inline core/src/cluster_info.rs:681]: maybe just 
```
let contact_info = v.value.contact_info().unwrap()
```
here.

File: core/src/cluster_info.rs [hunk 2]
```diff
@@ -653,6 +663,117 @@ impl ClusterInfo {
...
+        };
+
+        info!(
+            "Loaded contact info for {} nodes from {}",
+            nodes.len(),
+            filename.display()
+        );
+        let now = timestamp();
+        for node in nodes {
+            if let Err(err) = self.gossip.write().unwrap().crds.insert(node, now) {
```
Comment (behzadnouri) [inline core/src/cluster_info.rs:771]: maybe obtain this lock only once for the entire loop

**Review by behzadnouri — DISMISSED** [review]

File: core/src/cluster_info.rs (hunk 2 above)
Comment (mvines) [reply core/src/cluster_info.rs]: Ah, good idea. Done

File: core/src/cluster_info.rs (hunk 1 above)
Comment (mvines) [reply core/src/cluster_info.rs]: Ok.  Generally unwraps like this make me a little nervous, we've had crashes due to bad unwraps before.  But I guess in this case `get_nodes_contact_info()` already calls `get_nodes()` and explicitly panic on `None` so an unwrap here isn't any worse

Comment (behzadnouri) [reply core/src/cluster_info.rs]: Agree about avoiding unwrap in general, but in this case it is an invariant that *has to be always the case*.

**Pull request merged**